x-pack/libbeat/common/aws/signer.go

```diff
@@ -0,0 +1,195 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package aws
+
+import (
+	"bytes"
+	"crypto/sha256"
+	"encoding/hex"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/aws/aws-sdk-go-v2/aws"
+	v4 "github.com/aws/aws-sdk-go-v2/aws/signer/v4"
+	awslogging "github.com/aws/smithy-go/logging"
+
+	"github.com/elastic/elastic-agent-libs/logp"
+)
+
+type SignerInputConfig struct {
+	Enabled     *bool  `config:"enabled"`
+	ServiceName string `config:"service_name"` // optional
+	ConfigAWS   `config:",inline"`
+}
+
+// IsEnabled returns true if the `enable` field is set to true in the yaml.
+func (c *SignerInputConfig) IsEnabled() bool {
+	return c != nil && (c.Enabled == nil || *c.Enabled)
+}
+
+// SignerTransport implements [http.RoundTripper] interface
+// and sings requests with aws v4 signer before send them to the next roundtripper.
+type SignerTransport struct {
+	next        http.RoundTripper
+	credentials aws.CredentialsProvider
+	signer      *v4.Signer
+	logger      *logp.Logger
+	serviceName string
+	region      string
+	now         func() time.Time // we don't use [time.Now] directly, so we can mock time in tests.
+}
+
+func InitializeSingerTransport(cfg SignerInputConfig, logger *logp.Logger, nextTransport http.RoundTripper) (*SignerTransport, error) {
+	awsConfig, err := InitializeAWSConfig(cfg.ConfigAWS, logger)
+	if err != nil {
+		return nil, err
+	}
+
+	return initializeSingerTransport(logger, cfg.ServiceName, cfg.DefaultRegion, awsConfig.Credentials, nextTransport), nil
+}
+
+func initializeSingerTransport(logger *logp.Logger, defaultServiceName string, defaultRegion string, credentials aws.CredentialsProvider, nextTransport http.RoundTripper) *SignerTransport {
+	return &SignerTransport{
+		next:        nextTransport,
+		credentials: credentials,
+		signer: v4.NewSigner(func(signer *v4.SignerOptions) {
+			signer.Logger = awslogging.LoggerFunc(func(classification awslogging.Classification, format string, v ...any) {
+				switch classification {
+				case awslogging.Debug:
+					logger.Debugf(format, v...)
+				case awslogging.Warn:
+					logger.Warnf(format, v...)
+				}
+			})
+		}),
+		logger:      logger,
+		serviceName: defaultRegion,
+		region:      defaultServiceName,
+		now:         time.Now,
+	}
+}
+
+func (st *SignerTransport) RoundTrip(req *http.Request) (*http.Response, error) {
+	// resolve service name and region (if they are not configured)
+	serviceName, region, err := st.getServiceAndRegion(req)
+	if err != nil {
+		return nil, fmt.Errorf("error while getting service name and region: [%w]", err)
```

Comment:
```suggestion
		return nil, fmt.Errorf("error while getting service name and region: %w", err)
```
throughout (this is not conventional error syntax for Go)